[Calico] Fix delay setting up ip routes in new nodes

felipejfc opened the pull request:
Same as PR #4588 but with latest changes from master

mikesplain commented:
This is awesome, thanks for your work on this @felipejfc.

/ok-to-test
/assign

mikesplain commented:
We've been dealing with this for awhile too so I'll give this a test as well.

robinpercy commented on upup/pkg/fi/cloudup/bootstrapchannelbuilder.go:
```diff
@@ -469,8 +469,8 @@ func (b *BootstrapChannelBuilder) buildManifest() (*channelsapi.Addons, map[stri
 			"pre-k8s-1.6": "2.4.2-kops.1",
-			"k8s-1.6":     "2.6.7-kops.1",
-			"k8s-1.7":     "2.6.7-kops.1",
+			"k8s-1.6":     "2.6.8-kops.1",
```
We just need to update the suffix here: (ie 2.6.7-kops.2)

felipejfc replied:
the reason that I bumped the version to 2.6.8 is that if I didn't, then existing clusters would not get the current deployment and daemonset updated

mikesplain replied:
Right but `2.6.8` would imply a different image which is not the case, just a config change, right? I agree with @robinpercy.  We shouldn't need to bump the version, just the `kops.1 -> kops.2` number

felipejfc replied:
oh, got it!

mikesplain commented:
/assign robinpercy

felipejfc commented:
fixed the version @robinpercy @mikesplain

mikesplain commented:
Awesome! Thanks @felipejfc, I'm going to test this locally now!

robinpercy commented:
Thanks @felipejfc!
/lgtm

mikesplain commented:
I'm having issues testing this at the moment, will circle back tomorrow, but I think this should solve my issue too.  Thanks so much for the contribution!

/lgtm

mikesplain commented:
I was able to get a test through of this and it worked great! Thanks so much @felipejfc and @caseydavenport!

robinpercy commented:
/assign @chrislovecnm

chrislovecnm commented on docs/networking.md:
```diff
@@ -191,6 +191,15 @@ For help with Calico or to report any issues:
+#### Calico troubleshooting
+##### New nodes are taking minutes for syncing ip routes and new pods on them can't reach kubedns
+This is caused by nodes in the Calico etcd nodestore no longer existing. Due to the ephemeral nature of AWS EC2 instances, new nodes are brought up with different hostnames, and nodes that are taken offline remain in the Calico nodestore. This is unlike most datacentre deployments where the hostnames are mostly static in a cluster. Read more about this issue at https://github.com/kubernetes/kops/ ...
+This has been solved in kops 1.8.2, when creating a new cluster no action is needed, but if the cluster was created with a prior kops version the following actions should be taken:
```
This will need to be update to kops 1.9.0

felipejfc replied:
ok!

chrislovecnm commented on docs/networking.md:
````diff
@@ -191,6 +191,15 @@ For help with Calico or to report any issues:
+This is caused by nodes in the Calico etcd nodestore no longer existing. Due to the ephemeral nature of AWS EC2 instances, new nodes are brought up with different hostnames, and nodes that are taken offline remain in the Calico nodestore. This is unlike most datacentre deployments where the hostnames are mostly static in a cluster. Read more about this issue at https://github.com/kubernetes/kops/ ...
+This has been solved in kops 1.8.2, when creating a new cluster no action is needed, but if the cluster was created with a prior kops version the following actions should be taken:
+  * Use kops to update the cluster ```kops update cluster <name> --yes``` and wait for calico-kube-controllers deployment and calico-node daemonset to be updated
+  * Delete all calico-node pods in kube-system namespace, so that they will be recreated with the new env CALICO_K8S_NODE_REF and update the current nodes in etcd
````
Can we give kubectl command?

felipejfc replied:
I did it manually and right now I have no time thinking of a command that would do everything :/, maybe @mikesplain or @caseydavenport  have a script for that?

mikesplain replied:
Actually, I would suggest just rolling over the whole cluster... Deleting the pods could create temporary connectivity loss couldn't it? We drain and roll the cluster anytime we need to change core daemonsets like this.

felipejfc replied:
not actually, I always do ```kubectl delete pods -n kube-system -l "k8s-app=calico-node" --force --grace-period=0```
deleting calico-node pods does not delete the iptables rules and interfaces that were already created, so no problem on that.

felipejfc replied:
@chrislovecnm you mean kubectl command to delete calico-node pods or to delete the invalid nodes in etcd storage, the later is the one I have no time to provide, the first one is in my comment above

mikesplain replied:
@felipejfc Fair enough! I wasn't certain. I have a script/container open sourced that we could use: https://github.com/mikesplain/calico-clean/blob/master/calico-clean.sh

Would just need to convert https://github.com/mikesplain/calico-clean/blob/master/CronJob.yaml into a one time job.

caseydavenport replied:
Do we need to delete the calico-node pods? Won't upgrade the cluster trigger a rolling update due to the env change?

If not couldn't we enable rolling update on the daemonset and then it would?

Deleting all workload pods shouldn't be necessary (nor should rolling the whole cluster).

felipejfc replied:
I aggre it should be ok to add a rolling update strategy @caseydavenport, what do you guys think @mikesplain @chrislovecnm @robinpercy

mikesplain replied:
Personally, I've never used rolling update strategy on daemonset.  I could see cases when upgrading an entire cluster would trigger a daemonset upgrade which may not be intended on older nodes... We could put it behind a config value or something.  I have mixed feelings... Thoughts @chrislovecnm @robinpercy ?

caseydavenport replied:
It should only trigger when changes are made to the DaemonSet (not other things in the cluster) so I suspect it's OK and desirable.

We enable it by default in upstream Calico manifests.

chrislovecnm commented on docs/networking.md:
````diff
@@ -191,6 +191,15 @@ For help with Calico or to report any issues:
+This has been solved in kops 1.8.2, when creating a new cluster no action is needed, but if the cluster was created with a prior kops version the following actions should be taken:
+  * Use kops to update the cluster ```kops update cluster <name> --yes``` and wait for calico-kube-controllers deployment and calico-node daemonset to be updated
+  * Delete all calico-node pods in kube-system namespace, so that they will be recreated with the new env CALICO_K8S_NODE_REF and update the current nodes in etcd
+  * Decommission all invalid nodes, [see here](https://docs.projectcalico.org/v2.6/usage/decommissioning-a-node)
````
Should we just recommend rolling the cluster?

felipejfc replied:
it will have a higher overhead rolling update the cluster, what do you think if I just point it as an alternative?

ese commented:
should it be ported also for canal? https://github.com/kubernetes/kops/blob/master/upup/models/cloudup/resources/addons/networking.projectcalico.org.canal/k8s-1.8.yaml.template

felipejfc commented:
I guess that the damage this does to clusters using Canal is smaller @ese, still I don't see why not port it...

mikesplain commented:
@felipejfc If you want to open the canal change, I'd make it a separate PR so we can get this one through, and likely that one will move quicker once we figure out a few of these details :)

felipejfc commented:
so, I've applied an updateStrategy to calico-node daemonset (as per @caseydavenport recommendation), removed the "delete calico-node pods" step from the docs as its no longer necessary and updated the "fixed on" version to 1.9.0

mikesplain commented:
Great! Thanks so much @felipejfc and @caseydavenport!

/lgtm

Mind taking a look again @chrislovecnm

chrislovecnm commented:
/lgtm

We need to reference the doc in the release notes.
